sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuColumnarBatchSerializer.scala

```diff
@@ -491,6 +494,115 @@ private class KudoSerializerInstance(
     throw new UnsupportedOperationException
 }
 
+
+
+
+private class KudoGpuSerializerInstance(
+    val metrics: Map[String, GpuMetric],
+    val dataTypes: Array[DataType],
+) extends SerializerInstance {
+  private val serTime = metrics(METRIC_SHUFFLE_SER_STREAM_TIME)
+  private val deserTime = metrics(METRIC_SHUFFLE_DESER_STREAM_TIME)
+
+  override def serializeStream(out: OutputStream): SerializationStream = new SerializationStream {
+
+    override def writeValue[T: ClassTag](value: T): SerializationStream = serTime.ns {
+      val batch = value.asInstanceOf[ColumnarBatch]
+      val numColumns = batch.numCols()
+      withResource(new ArrayBuffer[AutoCloseable](numColumns)) { toClose =>
+        if (batch.numCols() > 0) {
+          val firstCol = batch.column(0)
+          // TODO: handle secondCol
+          firstCol match {
+            case vector: SlicedGpuCompressedColumnVector =>
+              withResource(vector.getWrap.getTableBuffer) { buf =>
+                val start = vector.getStart
+                val len = vector.getEnd - start
+                val data = buf.slice(start, len)
+                val hostBuffer = HostMemoryBuffer.allocate(data.getLength)
+                hostBuffer.copyFromDeviceBuffer(data);
```

Comment:
I personally would prefer to have us copy the data from the device to the host when the batch is serialized. That way we can release the GPU semaphore and know that we are off of the GPU.

It also removes the chance of us running out of host memory and needing to retry and allocation in multiple places.